pkg/compiler/types.go

```diff
@@ -81,15 +84,44 @@ var typeInt = &typeDesc{
 	Check: func(comp *compiler, t *ast.Type, args []*ast.Type, base prog.IntTypeCommon) {
 		typeArgBase.Type.Check(comp, t)
 	},
+	CheckConsts: func(comp *compiler, t *ast.Type, args []*ast.Type, base prog.IntTypeCommon) {
+		if len(args) > 0 && len(args[0].Colon) != 0 {
+			begin := args[0].Value
+			end := args[0].Colon[0].Value
+			if len(args) > 1 && begin == 0 && int64(end) == -1 {
+				// intN[0:-1, align] is a special value for 'all possible values',
+				// but aligned.
+				return
+			}
+			if end-begin > 1<<64-1<<32 {
+				comp.error(args[0].Pos, "bad int range [%v:%v]", begin, end)
+				return
+			}
+			size, _ := comp.parseIntType(t.Ident)
+			size = size * 8
+			if size < 64 && (end+1<<size > 2*(1<<size)-1 || begin+1<<size > 2*(1<<size)-1) {
```

Comment:
To be honest, I'm a little surprised the above works and the compiler doesn't try to optimize it. The idea is we want values in `[0:MAX_INTN]` or `[MAX_INT64-MAX_INTN:MAX_INT64]`. So we first shift all values by `MAX_INTN+1` and check bounds afterward.

Reply:
It's not totally clear to me what values make and don't make sense. I currently allow `intN[-MAX_UINTN:MAX_UINTN]`, but maybe we should only allow `intN[-MIN_SINTN:MAX_SINTN]` and `intN[0:MAX_UINTN]` (as I understand, `int32` can either mean `int32_t` or `uint32_t`)?

Reply:
I implemented a check for the above, which therefore prohibits `int8[-200:200]`, but also `int8[-200:0]`.